code/modules/clothing/glasses/scanners.dm

```diff
@@ -70,57 +58,112 @@
-			var/mob/living/carbon/monkey/M = C
-			if(M.glasses && (M.glasses == src))
-				remove_color(M)
-		else if(istype(loc, /mob/living/carbon/human))
-			var/mob/living/carbon/human/H = C
-			if(H.glasses && (H.glasses == src))
-				remove_color(H)
 
 /obj/item/clothing/glasses/scanner/night
 	name = "night vision goggles"
 	desc = "You can totally see in the dark now!"
 	icon_state = "night"
 	item_state = "glasses"
 	origin_tech = Tc_MAGNETS + "=2"
-	see_invisible = SEE_INVISIBLE_OBSERVER_NOLIGHTING
+	see_invisible = 0
+	seedarkness = TRUE
 	see_in_dark = 8
 	actions_types = list(/datum/action/item_action/toggle_goggles)
 	species_fit = list(VOX_SHAPED, GREY_SHAPED)
 	eyeprot = -1
-	color_matrix = list(0.8, 0, 0  ,\
-						0  , 1, 0  ,\
-						0  , 0, 0.8) //equivalent to #CCFFCC
-	my_dark_plane_alpha_override = "night_vision"
-	my_dark_plane_alpha_override_value = 255
+	var/obj/abstract/screen/plane_master/overdark_planemaster/overdark_planemaster
+	var/obj/abstract/screen/plane_master/overdark_planemaster_target/overdark_target
+
+/obj/item/clothing/glasses/scanner/night/New()
+	..()
+	overdark_planemaster = new
+	overdark_planemaster.render_target = "night vision goggles (\ref[src])"
+	overdark_target = new
+	overdark_target.render_source = "night vision goggles (\ref[src])"
 
 /obj/item/clothing/glasses/scanner/night/enable(var/mob/C)
-	see_invisible = initial(see_invisible)
 	see_in_dark = initial(see_in_dark)
 	eyeprot = initial(eyeprot)
+	my_dark_plane_alpha_override = "night_vision"
+	if (ishuman(C))
+		var/mob/living/carbon/human/H = C
+		if (H.glasses == src)
+			C.update_perception()
+			add_overdark(C)
+	else if (ismonkey(C))
+		var/mob/living/carbon/monkey/M = C
+		if (M.glasses == src)
+			C.update_perception()
+			add_overdark(C)
 	return ..()
 
 /obj/item/clothing/glasses/scanner/night/disable(var/mob/C)
 	. = ..()
-	see_invisible = 0
 	see_in_dark = 0
+	my_dark_plane_alpha_override = null
 	eyeprot = 0
+	remove_overdark(C)
+	if (ishuman(C))
+		var/mob/living/carbon/human/H = C
+		if (H.glasses == src)
+			if (C.client)
+				C.client.color = null
+			C.update_perception()
+	else if (ismonkey(C))
+		var/mob/living/carbon/monkey/M = C
+		if (M.glasses == src)
+			if (C.client)
+				C.client.color = null
+			C.update_perception()
+
+/obj/item/clothing/glasses/scanner/night/update_perception(var/mob/living/carbon/human/M)
+	if (on)
+		if (M.master_plane)
+			M.master_plane.blend_mode = BLEND_ADD
+		if (M.client)
+			M.client.color = "#33FF33"
+			remove_overdark(M)
+			add_overdark(M)
+	else
+		my_dark_plane_alpha_override = null
+		if (M.master_plane)
+			M.master_plane.blend_mode = BLEND_MULTIPLY
+
+/obj/item/clothing/glasses/scanner/night/equipped(var/mob/M, glasses)
+	if(istype(M, /mob/living/carbon/monkey))
+		var/mob/living/carbon/monkey/O = M
+		if(O.glasses != src)
+			return
+	else if(istype(M, /mob/living/carbon/human))
+		var/mob/living/carbon/human/H = M
+		if(H.glasses != src)
+			return
+	else
+		return
+	..()
```

Comment:
all above branches force an earlier return, so this parent call can never be reached